JavaScript/Angular.md

```diff
@@ -0,0 +1,1551 @@
+
+### Our own library for working with forms!
+
+Since we mostly use reactive forms, we created a library which make things a bit easier. The library in question is [ngx-form-object](https://github.com/infinum/ngx-form-object). Check out the project's readme on GitHub to see how to use it.
+
+Some of the benefits of using it are:
+- Definition of form attributes on model-level using decorators
+- Support for single-values and relationships (Attribute, BelongsTo, HasMany)
+- Introduces `FormObject` and `FormStore` for decoupling of data and validation
+- Form generation based on model definition
+
+We definitely recommend trying out `ngx-form-object` (*shameful self-promotion*). If you decide to use reactive forms. We used it on many complex projects with huge forms and it helped us a lot!
+
+### Making your components work with forms transparently
+
+Another topic which should be covered when talking about forms is using your own components with both reactive and template-driven forms. To use your component with forms just like you would use `input` element, you will have to implement [ControlValueAccessor](https://angular.io/api/forms/ControlValueAccessor) in your component. It isn't too straight-forward and might seem complex at first glance (what is `forwardRef` anyway, right?), but hopefully there are some great articles which we recommend reading:
+- [Angular Custom Form Controls Made Easy](https://netbasal.com/angular-custom-form-controls-made-easy-4f963341c8e2)
+  - quick introduction
+- [Never again be confused when implementing ControlValueAccessor in Angular forms](https://blog.angularindepth.com/never-again-be-confused-when-implementing-controlvalueaccessor-in-angular-forms-93b9eee9ee83)
+  - more in-depth, great article, recommend reading; but uses jQuery in example - so démodé
+
+You can implement both two-way binding and `ControlValueAccessor` if you want your component to be usable both inside a form and standalone outside the form, but usually you will not use the same component both inside and outside a form, so implementing just `ControlValueAccessor` is usually enough.
+
+## Testing
+
+Depending on project and client's requirements, you might not have to do any testing at all or you might have to target >90% coverage. Usefulness of this section will depend on your need and desire for testing.
+
+We do not recommend not wiring any tests and neither do we recommend mindlessly trying to achieve 100% coverage. What we do recommend is to consider writing good quality tests for critical functionalities. Applications with 100% coverage can still have bugs either because the test are not good enough or if edge cases were not covered by the implementation in the first place.
```

Comment:
```suggestion
We do not recommend not writing any tests and neither do we recommend mindlessly trying to achieve 100% coverage. What we do recommend is to consider writing good quality tests for critical functionalities. Applications with 100% coverage can still have bugs either because the test are not good enough or if edge cases were not covered by the implementation in the first place.
```